File: newton_square_root.pl
#!/usr/bin/perl
use strict;
use warnings;

sub square {
  my $x = shift;
  return $x * $x;
}

sub sq_root {
  my $x = shift;
  return sqrt_iter(1.0,$x);
}

sub sqrt_iter {
  my $guess = shift;
  my $x = shift;
  if (good_enough($guess,$x)) {
    return $guess;
  } else {
    return sqrt_iter(improve_guess($guess,$x),$x)
  }
}

sub improve_guess {
  my $guess = shift;
  my $x = shift;
  return average($guess,($x/$guess));
}

sub average {
  my $x = shift;
  my $y = shift;
  return ($x+$y)/2
}

sub good_enough {
  my $guess = shift;
  my $x = shift;
  return abs(square($guess) - $x) < 0.01;
}

my $one = sq_root(9);
print "$one\n";
my $two = sq_root(100+37);
print "$two\n";
my $three = sq_root(sq_root(2) + sq_root(3));
print "$three\n";
my $four = square(sq_root(1000));
print "$four\n";

File: newton_square_root.js
#!/usr/bin/env node
// newton's method for finding square roots

function square(x) { return x * x }

function sqrt(x) { return sqrtIter(1.0,x) }

function sqrtIter(guess,x) {
  if (goodEnough(guess,x)) {
    return guess;
  } else {
    return sqrtIter(improveGuess(guess,x),x);
  }
}

function improveGuess(guess,x) {
  return average(guess,x/guess)
}

function average(x,y) { return (x+y)/2 }

function goodEnough(guess,x) {
  return absolute(square(guess) - x) < 0.01
}

function absolute(x) {
  return x >= 0 ? x : -x;
}

console.log(sqrt(9));
console.log(sqrt(100+37));
console.log(sqrt(sqrt(2) + sqrt(3)));
console.log(square(sqrt(1000)));
